### PR TITLE
Argo Executor init container

### PR DESCRIPTION
**Pull request by sytianhe**
- Argo executor init container first commit
- Init container is responsible for downloading input artifacts for the user container
- Reading the Kubernetes downward api to interpret the workflow template (annotation file)
- Using minio go sdk to support private (minio), aws s3 and google api.

### CODE REVIEW

File: cmd/argoexec/commands/artifacts.go [hunk 1]
```diff
@@ -23,6 +39,131 @@ var artifactsLoadCmd = &cobra.Command{
...
+	// Read template
+	err := GetTemplateFromPodAnnotations(podAnnotationsPath, &wfTemplate)
+	if err != nil {
+		fmt.Printf("Error getting template %v\n", err)
+		os.Exit(1)
+	}
+
+	// Initialize in-cluster Kubernetes client
+	config, err := rest.InClusterConfig()
+	//config, err := clientcmd.BuildConfigFromFlags("", "/Users/Tianhe/.kube/cluster_minikube.conf")
```
Comment (jessesuen) [inline cmd/argoexec/commands/artifacts.go]: Remove personal settings. If you want, make kubeconfig as a flag for development purposes.

File: examples/input-artifact-s3.yaml [hunk 2]
```diff
@@ -11,9 +11,9 @@ spec:
       - name: CODE
         path: /src
         s3:
-          endpoint: https://storage.googleapis.com
-          bucket: my-bucket-name
-          key: path/in/bucket
+          endpoint: storage.googleapis.com
+          bucket: tianhe-test
```
Comment (jessesuen) [inline examples/input-artifact-s3.yaml]: Remove personal settings

File: examples/install.yaml [hunk 3]
```diff
@@ -4,43 +4,44 @@ metadata:
   name: workflow-controller-configmap
 data:
   config: |
-    executorImage: argoproj/argoexec:latest
+    executorImage: sytianhe/argoexec:latest
```
Comment (jessesuen) [inline examples/install.yaml]: Remove personal settings

File: cmd/argoexec/commands/artifacts.go [hunk 4]
```diff
@@ -23,6 +39,131 @@ var artifactsLoadCmd = &cobra.Command{
...
+			return nil
+		}
+	}
+
+	if err != io.EOF {
+		return errors.InternalWrapError(err)
+	}
+
+	// If reaching here, then no template prefix in the file
+	return errors.Errorf(errors.CodeInternal, "No template property found from annotation file: %s", annotationsPath)
```
Comment (jessesuen) [inline cmd/argoexec/commands/artifacts.go]: There is a convenience function for this: InternalErrorf

File: workflow/common/common.go [hunk 5]
```diff
@@ -27,6 +27,8 @@ const (
 	PodMetadataAnnotationsPath = PodMetadataMountPath + "/" + PodMetadataAnnotationsVolumePath
 	//
 	PodStatusVolumePath = "podstatus"
+	// PodAnnotationsTemplatePropertyName is the property name for template in the annotation file
+	PodAnnotationsTemplatePropertyName = "workflows.argoproj.io/template"
```
Comment (jessesuen) [inline workflow/common/common.go]: This is redundant to AnnotationKeyTemplate

**Comment by sytianhe** [comment]
Changed according to comments

**Pull request merged**